pkg/controller/controller.go

```diff
@@ -350,7 +353,9 @@ func (c *Controller) handleActionSet(as *crv1alpha1.ActionSet) (err error) {
 	if as, err = c.crClient.CrV1alpha1().ActionSets(as.GetNamespace()).Update(context.TODO(), as, v1.UpdateOptions{}); err != nil {
 		return errors.WithStack(err)
 	}
-	ctx := context.Background()
+	iv := getEnvAsIntOrDefault("ACTIONSET_TIMEOUT", 30)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(iv)*time.Second)
+	defer cancel()
 	ctx = field.Context(ctx, consts.ActionsetNameKey, as.GetName())
 	for i := range as.Status.Actions {
 		if err = c.runAction(ctx, as, i); err != nil {
```

Comment:
so here we are defining a context `ctx` and calling `runAction(ctx, as, i)`. `runAction()` has some synchronous execution, we only wait for that result in line 361. Apart from that `runAction()` method execute phases in a goroutine asynchronously(line 414) and we don't wait for that goroutine to finish. the goroutines also use the same context.
 So here the `handleActionSet()` completes the execution and calls `cancel()` on context even before asynchronous part of execution that uses the same context `ctx` completes. 
 cc @pavannd1